modules/calib3d/src/ap3p.cpp

```diff
@@ -0,0 +1,369 @@
+// solutionsT: 4 possible solutions of translation of the world origin w.r.t the camera frame
+int ap3p::computePoses(const double featureVectors[3][3],
+                       const double worldPoints[3][3],
+                       double solutionsR[4][3][3],
+                       double solutionsT[4][3]) {
+
+  //world point vectors
+  double w1[3] = {worldPoints[0][0], worldPoints[1][0], worldPoints[2][0]};
+  double w2[3] = {worldPoints[0][1], worldPoints[1][1], worldPoints[2][1]};
+  double w3[3] = {worldPoints[0][2], worldPoints[1][2], worldPoints[2][2]};
+  // k1
+  double u0[3];
+  vect_sub(w1, w2, u0);
+
+  double nu0 = vect_norm(u0);
+  double k1[3];
+  vect_divide(u0, nu0, k1);
+  // bi
+  double b1[3] = {featureVectors[0][0], featureVectors[1][0], featureVectors[2][0]};
+  double b2[3] = {featureVectors[0][1], featureVectors[1][1], featureVectors[2][1]};
+  double b3[3] = {featureVectors[0][2], featureVectors[1][2], featureVectors[2][2]};
+  // k3,tz
+  double k3[3];
+  vect_cross(b1, b2, k3);
+  double nk3 = vect_norm(k3);
+  vect_divide(k3, nk3, k3);
+
+  double tz[3];
+  vect_cross(b1, k3, tz);
+  // ui,vi
+  double v1[3];
+  vect_cross(b1, b3, v1);
+  double v2[3];
+  vect_cross(b2, b3, v2);
+
+  double u1[3];
+  vect_sub(w1, w3, u1);
+  // coefficients related terms
+  double u1k1 = vect_dot(u1, k1);
+  double k3b3 = vect_dot(k3, b3);
+  // f1i
+  double f11 = k3b3;
+  double f13 = vect_dot(k3, v1);
+  double f15 = -u1k1 * f11;
+  //delta
+  double nl[3];
+  vect_cross(u1, k1, nl);
+  double delta = vect_norm(nl);
+  vect_divide(nl, delta, nl);
+  f11 *= delta;
+  f13 *= delta;
+  // f2i
+  double u2k1 = u1k1 - nu0;
+  double f21 = vect_dot(tz, v2);
+  double f22 = nk3 * k3b3;
+  double f23 = vect_dot(k3, v2);
+  double f24 = u2k1 * f22;
+  double f25 = -u2k1 * f21;
+  f21 *= delta;
+  f22 *= delta;
+  f23 *= delta;
+  double g1 = f13 * f22;
+  double g2 = f13 * f25 - f15 * f23;
+  double g3 = f11 * f23 - f13 * f21;
+  double g4 = -f13 * f24;
+  double g5 = f11 * f22;
+  double g6 = f11 * f25 - f15 * f21;
+  double g7 = -f15 * f24;
+  double coeffs[5] = {g5 * g5 + g1 * g1 + g3 * g3,
+                      2 * (g5 * g6 + g1 * g2 + g3 * g4),
+                      g6 * g6 + 2 * g5 * g7 + g2 * g2 + g4 * g4 - g1 * g1 - g3 * g3,
+                      2 * (g6 * g7 - g1 * g2 - g3 * g4),
+                      g7 * g7 - g2 * g2 - g4 * g4};
+  double s[4];
+  solveQuartic(coeffs, s);
+  polishQuarticRoots(coeffs, s);
+
+  double temp[3];
+  vect_cross(k1, nl, temp);
+
+  double Ck1nl[3][3] =
+      {{k1[0], nl[0], temp[0]},
+       {k1[1], nl[1], temp[1]},
+       {k1[2], nl[2], temp[2]}};
+
+  double Cb1k3tzT[3][3] =
+      {{b1[0], b1[1], b1[2]},
+       {k3[0], k3[1], k3[2]},
+       {tz[0], tz[1], tz[2]}};
+
+  double b3p[3];
+  vect_scale((delta / k3b3), b3, b3p);
+
+  for (int i = 0; i < 4; ++i) {
+    double ctheta1p = s[i];
+    if (abs(ctheta1p) > 1)
+      continue;
```

Comment:
Here if the current `i` solution is not valid for some reason, I have experienced an issue when after the best solution is retained according to the reprojection error.